### PR TITLE
perf: make splitext ten times faster

### PR DESCRIPTION
**Pull request by radoering**
Since this is a hotspot during dependency resolution, this even makes "poetry lock" 10 % faster in some cases.

Incidentally, this also reverts the irrelevant change we ignored in https://github.com/python-poetry/poetry-core/pull/641/changes#r1460856881.

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: src/poetry/core/packages/utils/utils.py [hunk 1]
```diff
@@ -159,11 +160,28 @@ def is_archive_file(name: str | Path) -> bool:
...
+        In this case, we can use a performance shortcut
+        that only takes half of the time.
+    """
+    # Do not use pathlib because this method is a performance hotspot!
+    if is_filename:
+        # Using rpartition()
+        # - seems to be slightly faster than rsplit()
+        # - is twice as fast as os.path.splitext()
+        # - is by factor 10 faster than path.stem and path.suffix
+        base, sep, ext = str(path).rpartition(".")
```
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/core/packages/utils/utils.py:179]: These edge cases are not relevant for us.

File: tests/packages/utils/test_utils.py [hunk 2]
```diff
@@ -247,6 +248,25 @@ def test_get_python_constraint_from_marker(marker: str, constraint: str) -> None
...
+        ("demo-1.0.0.tar.gz", ("demo-1.0.0", ".tar.gz")),
+        ("demo-1.0.0.tar.bz2", ("demo-1.0.0", ".tar.bz2")),
+        ("demo-1.0.0.TAR.GZ", ("demo-1.0.0", ".TAR.GZ")),
+        ("dist/demo-1.0.0.tar.gz", ("dist/demo-1.0.0", ".tar.gz")),
+        (Path("demo-1.0.0.tar.gz"), ("demo-1.0.0", ".tar.gz")),
+        (Path("dist/demo-1.0.0.tar.gz"), (str(Path("dist/demo-1.0.0")), ".tar.gz")),
+    ],
+)
+def test_splitext(path: str | Path, expected: tuple[str, str]) -> None:
+    assert splitext(path) == expected
```
[the remark this answers is not in the document]
Comment (radoering) [reply tests/packages/utils/test_utils.py:272]: added tests

**Comment by radoering** [comment]
> * The `is_filename` flag makes `splitext`'s behavior depend on a correctness guarantee from the caller; consider adding a cheap runtime guard (e.g. assert that no path separator is present when `is_filename=True`) or splitting into two dedicated helpers to avoid subtle misuse later.

There probably is no "cheap runtime guard". The proposed assert is too expensive. The default is safe. Callers who set `is_filename=True` have to know what they are doing. Two dedicated helpers are probably not better than this.

File: src/poetry/core/packages/utils/link.py [hunk 3]
```diff
@@ -142,7 +142,7 @@ def path(self) -> str:
         return urlparse.unquote(urlparse.urlsplit(self.url)[2])
 
     def splitext(self) -> tuple[str, str]:
-        return splitext(posixpath.basename(self.path.rstrip("/")))
+        return splitext(posixpath.basename(self.path.rstrip("/")), is_filename=True)
```
Comment (abn) [inline src/poetry/core/packages/utils/link.py:145]: This collides with https://github.com/python-poetry/poetry-core/pull/941/changes#diff-2d216fd7b826c9ed25b2b621a4dc0281877ed80e3b38bd4d13a8ff97079c722cR155. Maybe best to leave this out here and let #941 override this.

**Review by abn — CHANGES_REQUESTED** [review]

File: src/poetry/core/packages/utils/link.py (hunk 3 above)
Comment (radoering) [reply src/poetry/core/packages/utils/link.py:145]: I am aware of the conflict, I will just resolve it after merging this one. Thus, both individual commits are correct.
